extensions/amp-analytics/0.1/transport.js

```diff
@@ -118,13 +152,220 @@ export class Transport {
     xhr.send('');
     return true;
   }
+
+  /**
+   * If iframe is specified in config/transport, check whether third-party
+   * iframe already exists, and if not, create it.
+   * @param {!Window} win The window element
+   * @param {!string} frameUrl  The URL of the cross-domain iframe
+   * @param {function(!string,
+   *   !../../../src/3p-analytics-common.AmpAnalytics3pResponse)=}
+   *   opt_processResponse An optional function to receive any response
+   *   messages back from the cross-domain iframe
+   */
+  processCrossDomainIframe(win, frameUrl, opt_processResponse) {
+    let frameData;
+    if (Transport.hasCrossDomainIframe(frameUrl)) {
+      frameData = Transport.getFrameData(frameUrl);
+      ++(frameData.usageCount);
+    } else {
+      frameData = this.createCrossDomainIframe(win, frameUrl);
+      win.document.body.appendChild(frameData.frame);
+    }
+    dev().assert(frameData, 'Trying to use non-existent frame');
+    if (opt_processResponse) {
+      Transport.responseProcessors_[this.id_] = opt_processResponse;
+    }
+  }
+
+  /**
+   * Create a cross-domain iframe for third-party vendor anaytlics
+   * @param {!Window} win The window element
+   * @param {!string} frameUrl  The URL of the cross-domain iframe
+   * @param {function(!string,
+   *   !../../../src/3p-analytics-common.AmpAnalytics3pResponse)=}
+   *   opt_processResponse An optional function to receive any response
+   *   messages back from the cross-domain iframe
+   * @return {!FrameData}
+   * @VisibleForTesting
+   */
+  createCrossDomainIframe(win, frameUrl, opt_processResponse) {
+    // Explanation of IDs:
+    // Each instance of Transport (owned by a specific amp-analytics tag, in
+    // turn owned by a specific creative) has an ID in this._id.
+    // Each cross-domain iframe also has an ID, stored here in sentinel.
+    // These two types of IDs are drawn from the same pool of numbers, and
+    // are thus mutually unique.
+    // There is a many-to-one relationship, in that several creatives may
+    // utilize the same analytics vendor, so perhaps creatives #1 & #2 might
+    // both use xframe #3.
+    // Of course, a given creative may use multiple analytics vendors, but
+    // in that case it would use multiple amp-analytics tags, so the
+    // transport.id_ -> sentinel relationship is *not* many-to-many.
+    const sentinel = Transport.createUniqueId_();
+    const useLocal = getMode().localDev || getMode().test;
+    const useRtvVersion = !useLocal;
+    const scriptSrc = calculateEntryPointScriptUrl(
+        window.location, 'ampanalytics-lib', useLocal, useRtvVersion);
+    const frameName = JSON.stringify(/** @type {JsonObject} */ ({
+      scriptSrc,
+      sentinel,
+    }));
+    const frame = createElementWithAttributes(win.document, 'iframe',
+        /** @type {!JsonObject} */ ({
+          sandbox: 'allow-scripts allow-same-origin',
```

Comment:
I banged my head against the wall for several hours figuring this out, but: if you don't do allow-same-origin, then it can't programmatically figure out what its origin is - see https://jsfiddle.net/keller_jon/L07ndhk6/ and then look at the console output.
So that in turn makes origin='null' in https://github.com/ampproject/amphtml/blob/master/src/iframe-helper.js and long story short, messages don't get delivered :(